File: apps/ios/Barns/BarnsApp/Domain/UseCases/Support/SaveConsultationDraftUseCase.swift
import Foundation

/// Saves the consultation draft locally. Local-only: the draft is never
/// sent to a server.
struct SaveConsultationDraftUseCase {
    private let repository: ConsultationDraftRepository

    init(repository: ConsultationDraftRepository) {
        self.repository = repository
    }

    func execute(
        existing: ConsultationDraft?,
        topic: String,
        category: ConsultationCategory,
        urgency: ConsultationUrgency,
        body: String
    ) async throws -> ConsultationDraft {
        let now = Date()
        let draft = ConsultationDraft(
            id: existing?.id ?? UUID().uuidString,
            productItemId: existing?.productItemId,
            topic: topic,
            category: category,
            urgency: urgency,
            body: body,
            status: .draft,
            createdAt: existing?.createdAt ?? now,
            updatedAt: now,
            imageUrl: nil
        )
        try await repository.saveDraft(draft)
        return draft
    }
}

File: apps/android/app/src/main/java/com/barns/app/domain/usecase/support/SaveConsultationDraftUseCase.kt
package com.barns.app.domain.usecase.support

import com.barns.app.domain.model.ConsultationCategory
import com.barns.app.domain.model.ConsultationDraft
import com.barns.app.domain.model.ConsultationDraftStatus
import com.barns.app.domain.model.ConsultationUrgency
import com.barns.app.domain.repository.ConsultationDraftRepository
import java.time.Instant
import java.util.UUID

/**
 * Saves the consultation draft locally. Local-only: the draft is never
 * sent to a server.
 */
class SaveConsultationDraftUseCase(
    private val repository: ConsultationDraftRepository,
) {
    suspend fun execute(
        existing: ConsultationDraft?,
        productItemId: String? = null,
        topic: String,
        category: ConsultationCategory,
        urgency: ConsultationUrgency,
        body: String,
    ): ConsultationDraft {
        val now = Instant.now()
        val draft = ConsultationDraft(
            id = existing?.id ?: UUID.randomUUID().toString(),
            productItemId = existing?.productItemId ?: productItemId,
            topic = topic,
            category = category,
            urgency = urgency,
            body = body,
            status = ConsultationDraftStatus.DRAFT,
            createdAt = existing?.createdAt ?: now,
            updatedAt = now,
            imageUrl = null,
        )
        repository.saveDraft(draft)
        return draft
    }
}
